src/lib/core/trigger.cc

```diff
@@ -54,24 +53,129 @@ trigger_fiber_f(va_list ap)
 	return trigger->run(trigger, event);
 }
 
+/**
+ * A single link in a list of triggers scheduled for execution. Can't be part of
+ * struct trigger, because one trigger might be part of an unlimited number of
+ * run lists.
+ */
+struct run_link {
+	/** A link in the run list belonging to one trigger_run() call.  */
+	struct rlist in_run;
+	/**
+	 * A link in a list of all links referencing that trigger. Used to
+	 * remove the trigger from whatever list it's on during trigger_clear().
+	 */
+	struct rlist in_trigger;
+	/** The trigger to be executed. */
+	struct trigger *trigger;
+};
+
+/** A memory pool for trigger run link allocation. */
+static __thread struct mempool run_link_pool;
+
+/** Add the trigger to the run list. */
+static int
+run_list_put_trigger(struct rlist *list, struct trigger *trigger)
+{
+	struct run_link *run_link =
+		(struct run_link *)mempool_alloc(&run_link_pool);
+	if (run_link == NULL) {
+		diag_set(OutOfMemory, sizeof(struct run_link),
+			 "mempool_alloc", "trigger run link");
+		return -1;
+	}
+	run_link->trigger = trigger;
+	rlist_add_tail_entry(list, run_link, in_run);
+	rlist_add_tail_entry(&trigger->run_links, run_link, in_trigger);
+	return 0;
+}
+
+/** Take the next trigger from the run list and free the corresponding link. */
+static inline struct trigger *
+run_list_take_trigger(struct rlist *list)
+{
+	struct run_link *link = rlist_shift_entry(list, struct run_link,
+						  in_run);
+	struct trigger *trigger = link->trigger;
+	rlist_del_entry(link, in_trigger);
+	mempool_free(&run_link_pool, link);
+	return trigger;
+}
+
+/** Empty the run list and free all the links allocated for it. */
+static inline void
+run_list_clear(struct rlist *list)
+{
+	while (!rlist_empty(list))
+		run_list_take_trigger(list);
+}
+
+/** Execute the triggers in an order specified by \a list. */
+static inline int
```

Comment:
Nit: Please avoid using `inline` in C files. The compiler will probably ignore it anyway. If you really need a function to be inlined (obviously, not in this case) you should use `ALWAYS_INLINE`.

Reply:
Sure, but why though? Only because the compiler usually ignores it?

Reply:
Just a general rule: if something isn't necessary, better not do it.

Reply:
Ok, seems legit, thanks.
Done.